feat: 배송경로 이벤트 처리 및 경로추가, 조회 api

maengjjin opened the pull request:
## PR 내용 
- 배송 생성 후 배송 경로 자동 생성을 위해 Spring Event 기반 비동기 처리 도입
- 배송 경로 추가 API 구현
- 배송 경로 조회 API 구현

Commit:
refactor: 메서드명 get() -> find()로 통일

Commit:
feat: 주문id로 배송조회, 배송전체 조회 구현

Commit:
refactor: 메서드명 findDriver → findArrivalHubDrivers 변경

Commit:
Merge branch 'develop' into feat/73/delivery-root

Commit:
feat: 배송 수정 API 변경

- 배송 수정 시 상태 변경이 아닌 담당자 변경으로 로직 수정
- 배송 담당자 변경 관련 테스트 코드 추가

Commit:
refactor: 메서드명 변경 및 누락된 파일 추가

- 메서드명 변경에 따른 테스트 코드 수정
- 이전 커밋에서 누락된 파일 추가

Commit:
feat: 배송삭제 api 구현

Commit:
Merge branch 'develop' into feat/73/delivery-root

Commit:
feat: 업체 FeignClient 적용
- hub-service와 name이 충돌하여 contextId 별도 설정

Commit:
feat: 배송생성 API 수정

- RequestBody 필드 구조 수정
- 배송생성에 필요한 배송담당자 FeignClient 설정 추가
- 관련 테스트 코드 수정 및 보완

Commit:
Merge branch 'refs/heads/develop' into feat/73/delivery-root

Commit:
feat: 배송 담당자·업체 조회 로직 분리 및 예외 처리 개선

- DeliveryService에서 담당자/업체 조회 로직 분리하여 책임 명확화
- 여러 서비스에서 사용할 수 있는 담당자 선택 기능을 유틸 패키지로 이동
- 조회 실패에 대한 예외 코드 및 메시지 정비
- 외부 API 장애 식별을 위한 로그 추가

Commit:
feat: 구 배송 수정 API 삭제 (신규 API로 대체)

- 기존 updateDelivery API 제거
- 신규 수정 API로 로직 전환 완료
- 에러코드 추가

Commit:
feat: 배송 등록 후 배송경로 생성 이벤트 처리 구현

- 배송 등록 완료 시 경로 생성을 위한 이벤트 발행 기능 추가
- 경로 생성 로직은 비동기(@Async)로 처리하여 응답 지연 최소화
- 중간 허브 배송 담당자 지정을 위해 FeignClient 연동 구현

Commit:
refactor: Delivery 고객 노출 배송 상태 매핑 규칙 정리

- 내부 Delivery 상태 → 고객용 상태(CREATED/SHIPPING/ARRIVED) 매핑 규칙 정리
- 고객이 이해하기 쉽도록 배송 상태를 간결하게 노출하도록 구조 개선
- 주석 및 띄어쓰기 정리로 가독성 향상

Commit:
feat: 배송경로 생성 기능 추가

- 배송경로 생성 기능 구현

Commit:
feat: 배송경로 조회 api 생성

- 경로조회 api 생성
- 배송경로 생성 시에도 경로 조회를 사용하여 따로 분리하여 경로 조회 api readonly로 변경

Commit:
refactor: DeliveryStatus → DeliveryRouteStatus로 이름 변경

Commit:
test: 배송경로 생성 시 상태 전환 및 배송 생성 테스트 보완

- 배송경로 생성 기능 테스트 추가
- 배송 생성 흐름 테스트 수정 및 강화

Commit:
Merge branch 'develop' into feat/97/delivery-event

# Conflicts:
#	delivery/src/main/java/com/klp/delivery/common/exception/ExternalApiErrorCode.java
#	delivery/src/main/java/com/klp/delivery/delivery/application/facade/DeliveryFacade.java
#	delivery/src/main/java/com/klp/delivery/delivery/application/service/DeliveryService.java
#	delivery/src/main/java/com/klp/delivery/delivery/application/service/DriverClientService.java
#	delivery/src/main/java/com/klp/delivery/delivery/domain/entity/Delivery.java
#	delivery/src/main/java/com/klp/delivery/delivery/exception/DeliveryErrorCode.java
#	delivery/src/main/java/com/klp/delivery/delivery/infrastructure/client/DriverClientServiceServiceImpl.java
#	delivery/src/main/java/com/klp/delivery/delivery/infrastructure/client/DriverFeignClient.java
#	delivery/src/main/java/com/klp/delivery/delivery/presentation/controller/DeliveryController.java
#	delivery/src/test/java/com/klp/delivery/delivery/application/DeliveryServiceTest.java
#	delivery/src/test/java/com/klp/delivery/delivery/application/facade/DeliveryFacadeTest.java
#	delivery/src/test/java/com/klp/delivery/delivery/presentation/controller/DeliveryIntegrationTest.java

dnjsals45 commented on delivery/src/main/java/com/klp/delivery/delivery/domain/entity/Delivery.java:
```diff
@@ -159,10 +155,15 @@ public void updateVendorDriverId(Long newVendorDriverId) {
     public void delete(Long deletedBy) {
 
-        if (this.status != DeliveryStatus.CREATED) {
+        if (this.status != CustomerDeliveryStatus.CREATED) {
```
이 부분도 위에 `updateVendorDriverId` 로직처럼
```java
public void delete(Long deltedBy) {
   if(!canDelete()) {}
}
```
이런식으로 변경되는게 더 좋아보일 것 같습니다

maengjjin replied:
넵 수정하겠습니다 :)

dnjsals45 commented on delivery/src/main/java/com/klp/delivery/delivery/domain/entity/DeliveryRoute.java:
```diff
@@ -0,0 +1,93 @@
+    @Column(name = "delivery_id")
+    private UUID deliveryId;
+
+    @Comment("업체 배송담당자 ID")
+    @Column(name = "drvier_id", nullable = false)
+    private Long driverId;
```
저희 플로우차트에서 변경사항에 따라 `고객 배송담당자` 로 바뀌면 좋을것 같아용

maengjjin replied:
플로우차트 변경이 있어 별도 브랜치를 만들어서 적용할려고 합니다!

dnjsals45 commented on delivery/src/main/java/com/klp/delivery/delivery/presentation/controller/DliveryRouteController.java:
```diff
@@ -0,0 +1,68 @@
+                route.getEstimatedTime(),
+                route.getRealDistance(),
+                route.getRealTime(),
+                route.getStatus()
+            ))
+            .toList();
```
컨트롤러에서 변환 작업을 하기보다는 dto 내부에서 static 메소드로 작성하면 더 깔끔하고, 책임 분리도 잘 될 것 같아요

maengjjin replied:
넵 수정하겠습니다 :)

dnjsals45 commented on delivery/src/main/java/com/klp/delivery/global/config/AsynConfig.java:
```diff
@@ -0,0 +1,34 @@
+    @Bean(name = "DeliveryRouteExecutor")
+    public ThreadPoolTaskExecutor paymentExecutor(){
+        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+        executor.setCorePoolSize(8);
+        executor.setMaxPoolSize(16);
+        executor.setQueueCapacity(500);
```
쓰레드 풀 사이즈, 큐 Capacity 등에 대해서는 설정 조건 등에 대해서는 이렇게 수치를 설정하신 이유가 있을까요? 이 쪽 관련 학습이 아직 부족해서 궁금해서 질문드립니다

dnjsals45 commented on delivery/src/main/java/com/klp/delivery/delivery/application/facade/DeliveryFacade.java:
```diff
@@ -110,11 +120,31 @@ private List<DeliveryResponse.DeliveryItemResponse> createDeliveriesForOrderItem
+        if (driver == null) {
+            throw new BusinessException(
+                DeliveryErrorCode.DELIVERY_CANNOT_BE_MODIFIED, "배송 담당자를 찾을 수 없습니다");
+        }
```
`driverService.findDriverAtArrivalHub()` 메소드에서 오류 발생 시 `DriverFeignClientErrorDecoder`에 의해서 에러가 처리될 것 같은데, driver가 null일 수 있는 가능성이 존재하나요?

maengjjin replied:
Feign 호출 시 orElseThrow 로 감싸져 있어 별도의 null 체크는 필요 없을것 같습니다 수정 하겠습니다 :)

dnjsals45 commented on delivery/src/main/java/com/klp/delivery/delivery/application/service/CompanyService.java:
```diff
@@ -0,0 +1,28 @@
+            return CompanyCommand.of(companyClientService.findCompany(customerId));
+        } catch (BusinessException e) {
+            throw e;
+        } catch (Exception e) {
+            log.error("업체 조회 실패: {}", e.getMessage(), e);
+            throw new BusinessException(DeliveryErrorCode.EXTERNAL_API_ERROR, "업체 조회에 실패했습니다.");
+        }
```
모든 Exception을 처리하기보다는 ExternalException일 때를 한번 더 분리하는 게 좋아 보입니다. 외부 호출이 아니라 다른 에러가 발생할 가능성도 존재하는데 모든 예외를 잡아버리고 있는 것 같아서요

dnjsals45 commented on delivery/src/main/java/com/klp/delivery/delivery/application/service/DeliveryRouteService.java:
```diff
@@ -0,0 +1,304 @@
+                    () -> new BusinessException(DeliveryErrorCode.DELIVERY_ROUTE_FETCH_FAILED,
+                        "마지막 배송 경로를 찾을 수 없습니다."));
+
+            // 다음 sequence의 PlanItem 찾기
+            int nextSequence = lastRoute.getSequence() + 1;
+            GetRoutePlanDetailResponse.PlanItem nextPlanItem = routePlan.planItems().stream()
+                .filter(item -> item.sequence().equals(nextSequence))
+                .findFirst()
+                .orElseThrow(
+                    () -> new BusinessException(DeliveryErrorCode.DELIVERY_ROUTE_FETCH_FAILED,
+                        "다음 경로 계획을 찾을 수 없습니다. sequence=" + nextSequence));
```
이 구조에서, `마지막 배송 경로의 + 1` 된 시퀀스를 찾는데, 이 시퀀스는 존재하지 않고, 결국 Throw되서 더 이상 로직 진행이 되지 않을 것 같은데 제 생각이 맞을까요?
마지막 경로인데, `마지막 경로의 + 1` 시퀀스가 존재할 수 있나요?

maengjjin replied:
해당 부분은 현재 배송 경로(DeliveryRoute)의 마지막 sequence를 의미하며, RoutePlan의 마지막 sequence와는 다를 수 있습니다
DeliveryRoute 의 마지막 sequence가 RoutePlan의 마지막 sequence보다 작은 경우, +1 된 sequence 값은 RoutePlan 상에 존재 합니다
변수명을 명확하게 수정하겠습니다!

dnjsals45 commented on delivery/src/main/java/com/klp/delivery/delivery/application/service/DeliveryService.java:
```diff
@@ -88,56 +47,68 @@ public Delivery registerDelivery(DeliveryCommand command, List<OrderItemCommand>
-            throw new BusinessException(DeliveryErrorCode.DRIVER_NOT_FOUND, "담당자 조회 결과가 없습니다.");
+    public void applyRouteCreation(UUID deliveryId, UUID routePlanId, CustomerDeliveryStatus status) {
+        try {
+            Delivery delivery = findDelivery(deliveryId);
+            delivery.updateRouteInfo(routePlanId, status);
+        } catch (BusinessException e) {
+            throw e;
+        } catch (Exception e) {
+            log.error("배송 경로 정보 적용 실패: {}", e.getMessage(), e);
+            throw new BusinessException(DeliveryErrorCode.EXTERNAL_API_ERROR, "배송 경로 정보 적용에 실패했습니다.", e);
         }
```
1. 트랜잭션 어노테이션이 빠져 있습니다.
2. 예외 처리가 잘못 진행되고 있는 것 같아요. 우선 JPA에서 find를 실패했을 때 기술적인 요인이면 Exception을 타고 타고 넘어와서 
```java
catch (Exception e) {
            log.error("배송 경로 정보 적용 실패: {}", e.getMessage(), e);
            throw new BusinessException(DeliveryErrorCode.EXTERNAL_API_ERROR, "배송 경로 정보 적용에 실패했습니다.", e);
        }
```
이 부분에 도착할 것 같은데, BAD_GATEWAY 예외 코드가 아니라고 생각합니다. 그리고 GlobalExceptionHandler 처리가 되어있을에도 try-catch로 계속해서 메소드마다 나누시는 이유는 로깅을 위한걸까요?

maengjjin replied:
해당 메서드의 부모 메서드에 트랜잭션이 적용되어있어 따로 트랜잭션으로 분리하지 않았습니다!

dnjsals45 reviewed (approved):
고생하셨습니다

Commit:
refactor: 피드백 기반 코드 정리

- 예외 처리 일관성 확보를 위해 BusinessException 만 사용하도록 수정
- 오타 및 불필요한 코드 정리
- 리뷰 피드백 반영하여 코드 가독성 개선

Commit:
refactor: 테스트 코드 수정 (예외 코드 변경 반영)
